internal/controller/nicconfigurationtemplate_controller.go

```diff
@@ -43,26 +57,262 @@ type NicConfigurationTemplateReconciler struct {
+		return reconcile.Result{}, nil
+	}
+	log.Log.Info("Reconciling NicConfigurationTemplates")
+
+	templateList := &v1alpha1.NicConfigurationTemplateList{}
+	err := r.List(ctx, templateList)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	log.Log.V(2).Info("Listed templates", "templates", templateList.Items)
+
+	deviceList := &v1alpha1.NicDeviceList{}
+	err = r.List(ctx, deviceList)
+	if err != nil {
+		log.Log.Error(err, "Failed to list NicDevices")
+		return ctrl.Result{}, err
+	}
+	log.Log.V(2).Info("Listed devices", "devices", deviceList.Items)
+
+	nodeList := &v1.NodeList{}
+	err = r.List(ctx, nodeList)
+	if err != nil {
+		log.Log.Error(err, "Failed to list cluster nodes")
+		return ctrl.Result{}, err
+	}
+	log.Log.V(2).Info("Listed nodes", "nodes", nodeList.Items)
+
+	nodeMap := map[string]*v1.Node{}
+	for _, node := range nodeList.Items {
+		nodeMap[node.Name] = &node
+	}
+
+	templates := []*v1alpha1.NicConfigurationTemplate{}
+	for _, template := range templateList.Items {
+		templates = append(templates, &template)
+	}
+
+	for _, device := range deviceList.Items {
+		node, ok := nodeMap[device.Status.Node]
+		if !ok {
+			log.Log.Info("device doesn't match any node, skipping", "device", device.Name)
+			continue
+		}
+
+		var matchingTemplates []*v1alpha1.NicConfigurationTemplate
+
+		for _, template := range templates {
+			if !deviceMatchesSelectors(&device, template, node) {
+				err := r.dropDeviceFromStatus(ctx, device.Name, template)
+				if err != nil {
+					log.Log.Error(err, "Failed to drop device name from template status", "template", template)
+				}
+
+				continue
+			}
+
+			matchingTemplates = append(matchingTemplates, template)
+		}
+
+		if len(matchingTemplates) == 0 {
+			log.Log.V(2).Info("Device doesn't match any configuration template, resetting the spec", "device", device.Name)
+			device.Spec.Configuration = nil
+			err = r.Update(ctx, &device)
+			if err != nil {
+				log.Log.Error(err, "Failed to update device's spec", "device", device)
+				return ctrl.Result{}, err
+			}
+			continue
+		} else if len(matchingTemplates) > 1 {
+			for _, template := range matchingTemplates {
+				err := r.dropDeviceFromStatus(ctx, device.Name, template)
+				if err != nil {
+					log.Log.Error(err, "Failed to drop device name from template status", "template", template)
+				}
+			}
 
-	// TODO(user): your logic here
+			templateNames := []string{}
+			for _, template := range matchingTemplates {
+				templateNames = append(templateNames, template.Name)
+			}
+			joinedTemplateNames := strings.Join(templateNames, ",")
+			err = fmt.Errorf("device matches several configuration templates: %s, %s", device.Name, joinedTemplateNames)
+			log.Log.Error(err, "Device matches several configuration templates, deleting its spec and emitting an error event", "device", device.Name)
+			return ctrl.Result{}, r.handleErrorSeveralMatchingTemplates(ctx, &device, joinedTemplateNames)
```

Comment:
why do we want to exit early and not continue to other devices ?
i would return only if error happened (then we reconcile again)

Reply:
You're right, not it only returns if an error occurred